### PR TITLE
Introduce `LauncherExecutionRequest`

### PR DESCRIPTION
**Pull request by marcphilipp**
## Overview

The new `Launcher.execute(LauncherExecutionRequest)` method paves the
road for adding additional parameters to test execution without having
to add additional overloads of `execute()`. For example, for #1880
we will likely need to introduce a `CancellationToken` (as drafted
in #4709). Instead of having to add two new overloads, such changes will
then be easy because only `LauncherExecutionRequest` will need to be
changed.

`LauncherExecutionRequestBuilder` provides a fluent API for constructing
execution requests starting either with a `TestPlan` or a
`LauncherDiscoveryRequest`. In addition,
`LauncherDiscoveryRequestBuilder.forExecution()` is a convenience method
to create an execution request from a discovery request.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://docs.junit.org/snapshot/user-guide/) and [Release Notes](https://docs.junit.org/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java [hunk 1]
```diff
@@ -323,6 +324,19 @@ else if (filter instanceof DiscoveryFilter<?> discoveryFilter) {
...
+	/**
+	 * Builds this discovery request and returns a new builder for creating a
+	 * {@link org.junit.platform.launcher.LauncherExecutionRequest} that is
+	 * initialized to contain the resulting discovery request.
+	 *
+	 * @return a new {@link LauncherExecutionRequestBuilder}
+	 * @since 6.0
+	 */
+	@API(status = EXPERIMENTAL, since = "6.0")
+	public LauncherExecutionRequestBuilder forExecution() {
```
Comment (sormuras) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java:336]:
```suggestion
	public LauncherExecutionRequestBuilder toLauncherExecutionRequestBuilder() {
```

Comment (sormuras) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java:336]: The following would be canonical, but also a mouthful:
```java
	public LauncherExecutionRequestBuilder toLauncherExecutionRequestBuilder() {
```

I think, that `forExecution` is a good-enough abbreviation here.

Comment (sormuras) [reply junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java:336]: Another variant could be:

```java
public LauncherExecutionRequest toLauncherExecutionRequest(UnaryOperator<LauncherExecutionRequestBuilder> operator) {
  return operator.apply(LauncherExecutionRequestBuilder.request(build())).build();
}
```

**Review by sormuras — APPROVED** [review]

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java (hunk 1 above)
Comment (marcphilipp) [reply junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java:336]: I've added this to our list of topics for the next team meeting. :+1:
